post_processing_stages/object_classify_tf_stage.cpp

```diff
@@ -0,0 +1,370 @@
+	threshold_high_ = params.get<float>("threshold_high", 0.01f);
+	threshold_low_ = params.get<float>("threshold_low", 0.01f);
+	model_file_ = params.get<std::string>("model_file", "/home/pi/models/mobilenet_quant_v1_224.tflite");
+	labels_file_ = params.get<std::string>("labels_file", "/home/pi/models/labels.txt");
+	display_labels_ = params.get<int>("display_labels", 1);
+
+	settings_.number_of_results = top_n_results_;
+	settings_.model_name = model_file_;
+	settings_.labels_file_name = labels_file_;
+
+	initialize();
+}
+
+void ObjectClassifyTfStage::initialize()
+{
+	model_ = tflite::FlatBufferModel::BuildFromFile(settings_.model_name.c_str());
+
+	settings_.model = model_.get();
+	std::cout << "Loaded model " << settings_.model_name << std::endl;
+	model_->error_reporter();
+
+	tflite::ops::builtin::BuiltinOpResolver resolver;
+
+	tflite::InterpreterBuilder(*model_, resolver)(&interpreter_);
+	if (!interpreter_)
+		throw std::runtime_error("Failed to construct interpreter");
+
+	std::cout << "Input size " << interpreter_->inputs().size() << std::endl;
+	interpreter_->SetAllowFp16PrecisionForFp32(settings_.allow_fp16);
+
+	if (settings_.number_of_threads != -1)
+		interpreter_->SetNumThreads(settings_.number_of_threads);
+
+	const std::vector<int> inputs_ = interpreter_->inputs();
+	const std::vector<int> outputs_ = interpreter_->outputs();
+
+	if (interpreter_->AllocateTensors() != kTfLiteOk)
+		throw std::runtime_error("Failed to allocate tensors");
+
+	if (readLabelsFile(settings_.labels_file_name, &labels_, &label_count_) != kTfLiteOk)
+		throw std::runtime_error("Failed to read labels file");
+}
+
+void ObjectClassifyTfStage::Configure()
+{
+	stream_ = app_->LoresStream();
+	if (stream_)
+	{
+		app_->StreamDimensions(stream_, &w_, &h_, &stride_);
+		if (w_ != IMAGE_WIDTH || h_ != IMAGE_HEIGHT)
+			throw std::runtime_error("Lores width and height must be 224");
+	}
+}
+
+bool ObjectClassifyTfStage::Process(CompletedRequest &completed_request)
+{
+	if (!stream_)
+		return false;
+
+	{
+		std::unique_lock<std::mutex> lck(future_ptr_mutex_);
+		if (completed_request.sequence % refresh_rate_ == 0 &&
+			(!future_ptr_ || future_ptr_->wait_for(std::chrono::seconds(0)) == std::future_status::ready))
+		{
+			libcamera::Span<uint8_t> buffer = app_->Mmap(completed_request.buffers[stream_])[0];
+			uint8_t *ptr = buffer.data();
+
+			tensor_input_ = yuvToRgb(ptr, w_, h_, stride_);
+
+			future_ptr_ = std::make_unique<std::future<void>>();
+			*future_ptr_ = std::move(std::async(std::launch::async, [this] {
+				auto time_taken =
+					ExecutionTime<std::micro>(std::bind(&ObjectClassifyTfStage::runInference, this)).count();
+				std::cout << time_taken << "ms" << std::endl;
+			}));
+		}
+	}
+
+	std::unique_lock<std::mutex> lock(output_results_mutex_);
+	std::string annotation = "Detected";
+
+	for (const auto &result : output_results_)
+	{
+		annotation = annotation + ", " + result.first + " : " + std::to_string(result.second);
+	}
```

Comment:
I think we could improve this a bit. If there's nothing detected we say "Detected: (nothing)", if there are several things "Detected: cat 73%, dog 61% or something like that without any extraneous commas or anything.